tests/run/carray_slicing.pyx

```diff
@@ -275,3 +276,227 @@ def struct_ptr_iter():
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+cpdef list test_coercing_c_slice_in_cpedef_list():
+    """
+    >>> test_coercing_c_slice_in_cpedef_list()
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+def test_c_to_py_coercing_array_in_var():
+    """
+    >>> test_c_to_py_coercing_array_in_var()
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    b = 5555
+    for i in range(10):
+        a[i] = i**2
+    b = a[8:]
+    return b
+
+cpdef list test_c_to_py_coercing_array_in_var_2():
+    """
+    >>> test_c_to_py_coercing_array_in_var_2()
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    b = 5555
+    for i in range(10):
+        a[i] = i**2
+    b = a[8:]
+    return b
+
+
+cpdef tuple test_coercing_c_slice_to_tuple():
+    """
+    >>> test_coercing_c_slice_to_tuple()
+    (64, 81)
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+cpdef set test_coercing_c_slice_to_set():
+    """
+    >>> b = test_coercing_c_slice_to_set()
+    >>> isinstance(b, set)
+    True
+    >>> b = list(b)
+    >>> b.sort()
+    >>> b
+    [64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+cpdef frozenset test_coercing_c_slice_to_frozenset():
+    """
+    >>> b = test_coercing_c_slice_to_frozenset()
+    >>> isinstance(b, frozenset)
+    True
+    >>> b = list(b)
+    >>> b.sort()
+    >>> b
+    [64, 81]
+    """
+
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+    return a[8:]
+
+
+def test_coercing_c_slice_by_casting():
+    """
+    >>> test_coercing_c_slice_by_casting()
+    [25, 36, 49, 64, 81]
+    (25, 36, 49, 64, 81)
+    [25, 36, 49, 64, 81]
+    [25, 36, 49, 64, 81]
+    """
+    cdef int i
+    cdef int a[10]
+    for i in range(10):
+        a[i] = i**2
+
+    print(<list> a[5:])
+    print(<tuple> a[5:])
+    b = <set> a[5:]
+    assert isinstance(b, set)
+    print(sorted(b))
+    b = <frozenset> a[5:]
+    assert isinstance(b, frozenset)
+    print(sorted(b))
+
+
+# Error !!!!!!!!
+#def test_coercing_c_slice_by_casting():
+#    """
+#    >>> test_coercing_c_slice_by_casting()
+#    [25, 36, 49, 64, 81]
+#    (25, 36, 49, 64, 81)
+#    set([64, 25, 36, 81, 49])
+#    frozenset([64, 25, 36, 81, 49])
+#    """
+#    cdef int i
+#    cdef int a[10]
+#    b = 5555
+#    for i in range(10):
+#        a[i] = i**2
+#    b = a
+#    print(<list> b)
+#    print(<tuple> b)
+#    print(<set> b)
+#    print(<frozenset> b)
+#
+#    # return lists
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+
+
+# Error !!!!!!!!
+#def test_coercing_c_slice_by_casting():
```

Comment:
It looks like you've fixed this (above)? Can this commented out bit now go? Or is this different in some way?

Reply:
These tests (commented) still fails
description of bug

b = 5555
# a can be c-array or slice of c-array
# b = a[1:5]
b = a

then when I am trying to cast

print(<list> b)
print(<tuple> b)
print(<set> b)
print(<frozenset> b)

result will be always the list, since dst_type will be always python object, not tuple_type, ... etc.


Second bug:
when b wasn't initialized before (like in previous code example)
b = a[1:5]
or
b = a
will raise the error

Reply:
The first bug looks like an existing problem that probably doesn't need fixing here.

I think the second bug is not really a bug. If `b` isn't initialized then Cython gets it's type from the assignment. You could also do `cdef object b` to set the type of `b`.